### PR TITLE
BHBC-1789: EML Updates + Survey Service

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## Links to Jira tickets

- {List all applicable Jira tickets}

## Description of relevant changes

- {List all relevant changes, in particular anything that will help the reviewers test/verify this PR}

## PR Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

File: api/src/paths/project/{projectId}/survey/permits/list.ts [hunk 1]
```diff
@@ -96,16 +97,19 @@ export function getSurveyPermits(): RequestHandler {
...
 
       await connection.commit();
 
-      const getSurveyPermitsData = (surveyPermitsData && surveyPermitsData.rows) || null;
+      const result = (response && response.rows) || null;
+
+      const getSurveyPermitsData = new GetPermitData(result);
+
+      if (!getSurveyPermitsData) {
+        return res.status(200).json(null);
```
Comment (KjartanE) [inline api/src/paths/project/{projectId}/survey/permits/list.ts:109]: Are you wanting to send 200 null back? Or just an empty array? If you are sending null back maybe we should update openApiSchema to allow nullable?

File: api/src/paths/project/{projectId}/surveys.ts [hunk 2]
```diff
@@ -53,35 +53,35 @@ GET.apiDoc = {
...
-                      type: 'string',
-                      description: 'ISO 8601 date string',
-                      nullable: true
-                    }
-                  }
-                },
-                species: {
-                  type: 'object'
-                }
+                // survey: {
```
Comment (KjartanE) [inline api/src/paths/project/{projectId}/surveys.ts]: Is this finished?

File: api/src/queries/survey/survey-view-queries.ts [hunk 3]
```diff
@@ -196,19 +196,49 @@ export const getSurveyFocalSpeciesDataForViewSQL = (surveyId: number): SQLStatem
   `;
 };
 
-export const getSurveyAncillarySpeciesDataForViewSQL = (surveyId: number): SQLStatement | null => {
+// export const getSurveyAncillarySpeciesDataForViewSQL = (surveyId: number): SQLStatement | null => {
```
Comment (KjartanE) [inline api/src/queries/survey/survey-view-queries.ts]: Deleting function?

File: api/src/services/project-service.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import moment from 'moment';
-import SQL from 'sql-template-strings';
+//import SQL from 'sql-template-strings';
```
Comment (KjartanE) [inline api/src/services/project-service.ts]: remove?

File: api/src/services/survey-service.ts [hunk 5]
```diff
@@ -67,11 +100,114 @@ export class SurveyService extends DBService {
...
+  async getSurveyPurposeAndMethodology(surveyId: number): Promise<object> {
+    const sqlStatement = queries.survey.getSurveyPurposeAndMethodologyForUpdateSQL(surveyId);
+
+    if (!sqlStatement) {
+      throw new HTTP400('Failed to build SQL get statement');
+    }
+
+    const response = await this.connection.query(sqlStatement.text, sqlStatement.values);
+
+    if (!response || !response?.rows?.[0]) {
```
Comment (KjartanE) [inline api/src/services/survey-service.ts]: Should the error check convention be continued here? Similar to above and below.

File: api/src/paths/project/{projectId}/surveys.ts (hunk 2 above)
Comment (anissa-agahchen) [reply api/src/paths/project/{projectId}/surveys.ts]: it is now, can you check again ?

File: api/src/queries/survey/survey-view-queries.ts (hunk 3 above)
Comment (anissa-agahchen) [reply api/src/queries/survey/survey-view-queries.ts]: Yes, in the getSpeciesData function, I now get the species list once from the DB and parse them into focal and ancillary using a constructor.

File: api/src/paths/project/{projectId}/survey/permits/list.ts (hunk 1 above)
Comment (anissa-agahchen) [reply api/src/paths/project/{projectId}/survey/permits/list.ts:109]: Thanks, good catch.   I made the permit section nullable.

**Comment by anissa-agahchen** [comment]
With this payload:
{"survey_details":{"survey_name":"Kasimir Townsend","start_date":"1900-05-01","end_date":"2064-10-02","biologist_first_name":"Chester","biologist_last_name":"Hopkins"},"species":{"focal_species":[1013,1188],"ancillary_species":[7233]},"permit":{"permit_number":"1","permit_type":""},"funding":{"funding_sources":[1]},"purpose_and_methodology":{"intended_outcome_id":5,"additional_details":"Dolores quo qui atqu","field_method_id":19,"ecological_season_id":7,"vantage_code_ids":[1],"surveyed_all_areas":"false"},"location":{"survey_area_name":"Danielle Franklin","geometry":[]},"proprietor":{"survey_data_proprietary":"false","proprietary_data_category":0,"proprietor_name":"","first_nations_id":0,"category_rationale":"","disa_required":"false"},"agreements":{"sedis_procedures_accepted":true,"foippa_requirements_accepted":true}},

the DB error is:

{"length":271,"name":"error","severity":"ERROR","code":"23502","detail":"Failing row contains (3, 3, 2, 1, 1, null, null, null, null, null, null, null, 2022-06-09 13:18:45.118499-07, 3, null, null, 0).","schema":"biohub","table":"permit","column":"type","file":"execMain.c","line":"1965","routine":"ExecConstraints","message":"null value in column \"type\" violates not-null constraint"}

**Comment by anissa-agahchen** [comment]
With this payload (no proprietor data):

{"length":271,"name":"error","severity":"ERROR","code":"23502","detail":"Failing row contains (3, 3, 2, 1, 1, null, null, null, null, null, null, null, 2022-06-09 13:18:45.118499-07, 3, null, null, 0).","schema":"biohub","table":"permit","column":"type","file":"execMain.c","line":"1965","routine":"ExecConstraints","message":"null value in column \"type\" violates not-null constraint"}

Error:

{"length":203,"name":"error","severity":"ERROR","code":"P0001","where":"PL/pgSQL function tr_survey_proprietor() line 33 at RAISE","file":"pl_exec.c","line":"3876","routine":"exec_stmt_raise","message":"The proprietor name is a required field when there is no associated first nations."}

File: Makefile [hunk 6]
```diff
@@ -100,15 +100,17 @@ build-web: ## Builds all backend+web containers
 	@echo "==============================================="
 	@echo "Make: build-web - building web images"
 	@echo "==============================================="
-	@docker-compose -f docker-compose.yml build db db_setup api app n8n n8n_nginx n8n_setup
+	@docker-compose -f docker-compose.yml build db db_setup api app
```
Comment (anissa-agahchen) [inline Makefile:103]: revert this back if you merge your PR before mine

**Comment by anissa-agahchen** [comment]
Editing any project section is now failing ... all related to fields required by the API response.

**Comment by anissa-agahchen** [comment]
I created a project with rectangular area, and the points showed up as expected:
<boundingCoordinates>              
                            <westBoundingCoordinate>-126.765656</westBoundingCoordinate>
                            <eastBoundingCoordinate>-124.449118</eastBoundingCoordinate>
                            <northBoundingCoordinate>56.079041</northBoundingCoordinate>
                            <southBoundingCoordinate>55.411321</southBoundingCoordinate>
                        </boundingCoordinates>
 but then my survey didn't have a geographic area, and the eml created:
  <boundingCoordinates>                
                                <westBoundingCoordinate>Infinity</westBoundingCoordinate>
                                <eastBoundingCoordinate>-Infinity</eastBoundingCoordinate>
                                <northBoundingCoordinate>-Infinity</northBoundingCoordinate>
                                <southBoundingCoordinate>Infinity</southBoundingCoordinate>
                            </boundingCoordinates>
   and the eml parser is not happy.

**Comment by anissa-agahchen** [comment]
Missing metadata fields:
In project:
* General Information: projectType
* General Information: timeline (start and end dates)
* GeneralInformation: activities
* Project contact: agency

In survey:
* General information : Survey Lead
* Intended Outcomes section
* Purpose and methodology section

**Comment by NickPhura** [comment]
> Editing any project section is now failing ... all related to fields required by the API response.

Should be fixed now.  The openapi spec just needed a slight tweak.

**Comment by NickPhura** [comment]
> I created a project with rectangular area, and the points showed up as expected: -126.765656 -124.449118 56.079041 55.411321 but then my survey didn't have a geographic area, and the eml created: Infinity -Infinity -Infinity Infinity and the eml parser is not happy.

Fixed.  If a survey doesn't have geometry, it just wont include that section now.

**Comment by NickPhura** [comment]
> Missing metadata fields: In project:
> 
>     * General Information: projectType
> 
>     * General Information: timeline (start and end dates)
> 
>     * GeneralInformation: activities
> 
>     * Project contact: agency
> 
> 
> In survey:
> 
>     * General information : Survey Lead
> 
>     * Intended Outcomes section
> 
>     * Purpose and methodology section

These fields should be accounted for now:
- project
  - project_type: added under additional metadata
  - timeline: was already there, under temporal coverage
  - activities: added under additional metadata
- survey
  - survey lead: added under personnel
  - intended outcomes/additional details: added under abstract
  - purpose/methodology: added under designDescription
  - surveyed all areas: added under additional metadata

**Comment by anissa-agahchen** [comment]
When I choose a single point on the map, I get this error:

XML specific tests: Failed
The following errors were found:

cvc-complex-type.2.4.j: The content of element 'datasetGPolygonOuterGRing' is not complete. 'gRingPoint' is expected to occur a minimum of '3' times. '2' more instances are required to satisfy this constraint.

**Comment by anissa-agahchen** [comment]
When I pick a line for the project boundary, I get:

EML Parser
EML specific tests: Failed.
The following errors were found:

null

XML specific tests: Failed
The following errors were found:

The processing instruction target matching "[xX][mM][lL]" is not allowed.

**Comment by anissa-agahchen** [comment]
vantage codes is pulling from the ecological seasons table
Screen Shot 2022-06-20 at 10 58 32 AM

**Comment by anissa-agahchen** [comment]
Data from the `Proprietary Data` section is missing:

* proprietary name
* data category
* disa required
* category rationale

**Comment by anissa-agahchen** [comment]
The sensitivedata flag is not making a difference to whether personnel data is shown or not.  I think the issue is that you're not passing the flag to the loadprojectdata and loadsurveydata functions.

**Comment by NickPhura** [comment]
> The sensitivedata flag is not making a difference to whether personnel data is shown or not. I think the issue is that you're not passing the flag to the loadprojectdata and loadsurveydata functions.

Just tried it and it seems to be working.

The `contact` and `personnel` sections will always appear.  The only thing that gets hidden is the persons name and their email inside that section. Organization will always show.

The `permit` additional metadata section will toggle completely though.

**Review by NickPhura — DISMISSED** [review]

**Review by NickPhura — APPROVED** [review]

**Pull request merged**